File: 백준/Gold/14719. 빗물/빗물.py
h, w = map(int, input().split())
blocks = list(map(int, input().split()))
result = 0

for i in range(1, w - 1):  #맨 왼쪽과 오른쪽은 고일 수 없음.
    left_max = max(blocks[:i])
    right_max = max(blocks[i + 1:])

    lower_one = min(left_max, right_max)
    if blocks[i] < lower_one:  #현재가 lower_one보다 낮아야 고일 수 있음
        result += lower_one - blocks[i]

print(result)

File: 백준/Gold/14719. 빗물/빗물.java
import java.util.*;
import java.io.*;

public class Main{

	public static void main(String[] args) throws IOException{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		
		int H = Integer.parseInt(st.nextToken()); // 높이
		int W = Integer.parseInt(st.nextToken());
		
		List<Integer> list = new ArrayList<>();
		st = new StringTokenizer(br.readLine());
		for(int i=0; i<W ; i++) {
			list.add(Integer.parseInt(st.nextToken()));
		}
		
		int rightMax =0, leftMax=0, quantity=0;
		for(int i=1; i<W-1; i++) {
			rightMax = Collections.max(list.subList(i+1,W));
			leftMax = Collections.max(list.subList(0,i));
			
			int lowerOne = Math.min(leftMax,rightMax);
			if(list.get(i)<lowerOne) {
				quantity += lowerOne-list.get(i);
			}
		}
		System.out.println(quantity);
		
	}

}
